### PR TITLE
Add missing actionFunc forward-declares

### PR DESCRIPTION
**Pull request by Kelebek1**
Before opening this PR, ensure the following:
- `./format.sh` was run to apply standard formatting.
- `make` successfully builds a matching ROM.
- No new compiler warnings were introduced during the build process.
    - Can be verified locally by running `tools/warnings_count/check_new_warnings.sh`
- New variables & functions should follow standard naming conventions.
- Comments and variables have correct spelling.
---

Most files had actionFuncs already forward-delcared, but some were missing. This PR tries to add the remaining ones, but also strips out functions which don't need to be forward declared. I left the Setup function in for actors with one, even though it doesn't need to be there in any actor I don't think. I could remove that too if needed.

### CODE REVIEW

**Comment by AngheloAlf** [comment]
I don't really see the benefit of this PR.
Considering we have way _too_ many open PRs right now, I think this will mainly produce a lot of conflicts (either by other people having to update their branches or you having to update this when another actor is merged). This will become a bit more of a headache if people rename the functions.

Also, what was your heuristic to remove forward declared functions? If they weren't needed to be forward declared as an action func? Are you sure those are never called as a normal function?

**Comment by Kelebek1** [comment]
Yeah, PRs shouldn't sit open for months. There's a lot which should just be accepted now no problem. I can't really do anything about that. This could be accepted already right now and then there's no conflicts.

There shouldn't be too many actively worked on open actors, so I don't think it'll be much of an issue, given that if they are that out-dated, they'll have to fix many merge issues anyway, this isn't a bad one.

The heuristic is if they're not needed, yeah. It also handles if they're normal functions.

**Comment by engineer124** [comment]
> Yeah, PRs shouldn't sit open for months. There's a lot which should just be accepted now no problem. I can't really do anything about that. This could be accepted already right now and then there's no conflicts.

I strongly disagree that PR's should just be accepted without proper review. (I don't have any strong opinions on the contents of this PR, but I did want to address that point)

**Comment by Kelebek1** [comment]
They do get reviewed. #465 , #463, #462, #454, #446 can be merged. Most of your PRs adding header stuff can be merged. Elliptic's new reloc program, Roman's cleanup, surpressing warnings. We have sooo many that could be added and don't need a month of reviews.

**Comment by AngheloAlf** [comment]
I agree PRs shouldn't sit down open for months, but it isn't a problem that would be solved by just merging everything as it comes in.
- People need time to properly look at each PR to see and understand what the changes are. Don't forget this is a project we do in our free time.
- It is necessary to properly look at the code because of any problem/typo/spelling mistake/etc the contributor in his contribution.
- There's also the problem of people not addressing the reviews of PRs they have made for weeks.
- We need to check nothing harmful is being PR'd.
- We also need to check there are no hidden changes in a PR that could affect more people. For example the change you made to m2ctx.py was part of an actor PR without any explanation or mention of what it was or why the change.
- Changes that touch _a lot_ of files like this one need to be carefully reviewed to check it is doing what it is stating to do. 
  - As you have said, the forward declarations we currently have may not be the best right now (because or missing or extra prototypes), but just PRing what seems to be just a lot of random changes doesn't seem to be the answer.
  - To properly review this PR we would need to pull this branch and check carefully the 200+ files you are changing.

> This could be accepted already right now and then there's no conflicts.

This isn't true, just merging this would produce conflicts with currently open PRs.

For the PRs you said we could just merge:
- #465 You and me discussed a lot of the changes you made to the disassembler the very first day it got PRd. I think more people may be interested in looking at what else changed.
- #463 Engineer catched an issue with that PR, it is possible there are more with it. The more people look at it, the better.
- #462 There still is discussion in that PR about the descriptions given to those actors. 

In other words, there no need to rush everything. Everything should be properly looked at before merging it. The more people reviewing everything, the better

**Comment by Thar0** [comment]
I disagree with removing (potentially) unnecessary forward declarations. The benefit for having them is for context generation and should be pruned by the one decompiling the actor imo

**Comment by EllipticEllipsis** [comment]
In addition to the comments made above, I would say that this PR is essentially un-reviewable: we can't verify that any of these changes are valid without going through essentially the same process as you did, which is much more difficult when you haven't explained in any detail how that worked.

**Comment by Kelebek1** [comment]
Ok, all actionFuncs are now preserved, regardless of whether they need declaring or not, so this now just adds the new ones.

**Comment by Kelebek1** [comment]
> In addition to the comments made above, I would say that this PR is essentially un-reviewable: we can't verify that any of these changes are valid without going through essentially the same process as you did, which is much more difficult when you haven't explained in any detail how that worked.

It gets the actionFunc offset from each actor, then scans the asm file for saves to that offset. It collects the function being saved to it and then prints them at the top.

**Comment by hensldm** [comment]
I do appreciate the work to try and make decomp easier by pre-providing all the actionfuncs, but I'm not to sure on the approach here. For instance in `EnFuMato`, it adds 5 new declarations. Throwing the file into `mips_to_c`, I can confirm one of these gets set to the variable `actionFunc`. The other 4 get set to some other variable, that doesn't seem used in the actor so impossible to tell what it should be. The prototype is probably right, but without context it is impossible to tell. IMO, it is better to leave it to the decomper since they can figure out the context.

**Comment by Kelebek1** [comment]
> I do appreciate the work to try and make decomp easier by pre-providing all the actionfuncs, but I'm not to sure on the approach here. For instance in `EnFuMato`, it adds 5 new declarations. Throwing the file into `mips_to_c`, I can confirm one of these gets set to the variable `actionFunc`. The other 4 get set to some other variable, that doesn't seem used in the actor so impossible to tell what it should be. The prototype is probably right, but without context it is impossible to tell. IMO, it is better to leave it to the decomper since they can figure out the context.

They're all saved to this->15C, which is actionFunc.

**Comment by hensldm** [comment]
> > I do appreciate the work to try and make decomp easier by pre-providing all the actionfuncs, but I'm not to sure on the approach here. For instance in `EnFuMato`, it adds 5 new declarations. Throwing the file into `mips_to_c`, I can confirm one of these gets set to the variable `actionFunc`. The other 4 get set to some other variable, that doesn't seem used in the actor so impossible to tell what it should be. The prototype is probably right, but without context it is impossible to tell. IMO, it is better to leave it to the decomper since they can figure out the context.
> 
> They're all saved to this->15C, which is actionFunc.

Oh you are right, I misread the mips_to_c output. Sorry about that.
